cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,494 @@
+				expBucketObjs, ok := testdata2.expectedObjects[bucket]
+				if !ok {
+					t.Errorf("bucket '%s' shouldn't exist in objects map", bucket)
+					continue
+				}
+
+				if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", bucket) {
+					continue
+				}
+
+				for expPath, expObj := range expBucketObjs {
+					if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+						continue
+					}
+
+					obj := bucketObjs[expPath]
+					assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+					assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+					assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+					assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+				}
+			}
+		}
+
+		// Assert that objserver keep track global stats of all the segments which
+		// have received through processSegment calls
+		assert.Equal(t,
+			testdata1.expectedInlineSegments+testdata2.expectedInlineSegments,
+			obsvr.inlineSegments,
+			"inlineSegments",
+		)
+		assert.Equal(t,
+			testdata1.expectedInlineSegments+testdata2.expectedInlineSegments,
+			obsvr.lastInlineSegments,
+			"lastInlineSegments",
+		)
+		assert.Equal(t,
+			testdata1.expectedRemoteSegments+testdata2.expectedRemoteSegments,
+			obsvr.remoteSegments,
+			"remoteSegments",
+		)
+	})
+
+	t.Run("object without last segment", func(t *testing.T) {
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		var (
+			testdata = generateTestdataObjects(ctx.Context, t, true, false)
+			obsvr    = observer{
+				objects: make(bucketsObjects),
+			}
+		)
+
+		// Call processSegment with the testdata
+		for _, objSeg := range testdata.objSegments {
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		// Assert observer internal state
+		assert.Equal(t, testdata.projectID.String(), obsvr.lastProjectID, "lastProjectID")
+		if assert.Equal(t, len(testdata.expectedObjects), len(obsvr.objects), "objects number") {
+			for bucket, bucketObjs := range obsvr.objects {
+				expBucketObjs, ok := testdata.expectedObjects[bucket]
+				if !ok {
+					t.Errorf("bucket '%s' shouldn't exist in objects map", bucket)
+					continue
+				}
+
+				if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", bucket) {
+					continue
+				}
+
+				for expPath, expObj := range expBucketObjs {
+					if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+						continue
+					}
+
+					obj := bucketObjs[expPath]
+					assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+					assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+					assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+					assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+				}
+			}
+		}
```

Comment:
These block is the same in at least 3 test cases. Can we extract it in a helper method?

Reply:
I'll take a look if they are completely equal.
I preferred to keep it where they are because from my point of view I prefer to keep what the test is checking in the test function. Anyway, when writing the test was easier to do in that way and tweak any assertion in case that needed for each individual test than starting having a helper function that later on I may have realized that I cannot homogenize.